engine/src/flutter/shell/platform/linux/fl_engine.cc

```diff
@@ -34,6 +34,9 @@ static constexpr size_t kPlatformTaskRunnerIdentifier = 1;
 static constexpr int32_t kMousePointerDeviceId = 0;
 static constexpr int32_t kPointerPanZoomDeviceId = 1;
 
+static int64_t next_handle = 1;
+static GHashTable* engine_map;
```

Comment:
Having these global values is a red flag - instead the handle should be set with `fl_engine_set_handle` and retrieved with `fl_engine_get_handle`. If we need to look up an engine it would probably live in the place that owns the engines, which is likely to be `FlApplication` (though we only support one engine at this point).

Reply:
The engine must be obtainable from handle without any other context. Requiring FlApplication instance in order to lookup the engine seems to negate the purpose of this PR?

Reply:
Reading the linked issue, shouldn't the handle just be the memory address of the engine, and then this can used by the FFI directly? Then we don't need any IDs or maps.

Reply:
I want to avoid casting it to pointer directly and also detect invalid / outdated handles, so that we can ensure that `fl_engine_for_handle` will return `null` if the engine no longer exists.

Reply:
Alternatively, if we relax the requirements, `fl_engine_for_handle` implementation can be 
```cpp
g_return_val_if_fail(FL_IS_ENGINE(engine_handle), nullptr);
return FL_ENGINE(engine_handle)
```
I don't feel too strongly about this, if the global map and counter is too much of an issue.

Reply:
On other platforms we don't have similar checks though, so we'll probably still be using maps. Not sure if inconsistent behavior between platforms is a good trade-off.

Reply:
The handle should remain valid until `FlutterEngineShutdown` is called. If the platform and UI thread is the same then we should always have safe interactions.

Reply:
Personally, I also like Robert's suggestion of using the engine's memory address.

> ... so that we can ensure that `fl_engine_for_handle` will return `null` if the engine no longer exists.

While this would be a nice check, it doesn't completely solve the problem of a plugin holding to an engine pointer after it is destroyed.

Windows already has APIs like `ViewControllerFromHandle` that will gladly return garbage if you pass a handle that no longer exists.

> Not sure if inconsistent behavior between platforms is a good trade-off.

We could claim the behavior is consistent by stating `fl_engine_for_handle` has undefined behavior if the handle no longer exists? 🙃

Reply:
`ViewControllerFromHandle` simply casts `FlutterDesktopViewControllerRef` (C API) to `flutter:FlutterWindowsViewController` (c++ class). I don't think that's the same. In think case the "handle" is clearly a pointer. I think better comparison would be that once we have the ViewForId method - that will never return garbage for invalid Id. 

I think it all depends on the threading mdoel. Passing raw engine pointer as handle can be dangerous if the plugin needs to do the hop from UI to platform thread, because while the callback waits to be executed on platform thread the engine might be destroyed and the plugin has have no way of detecting that handle is invalid. 

This would be much less of an issue if we get the threading change in.

Reply:
@robert-ancell, this is the implementation for `fl_engine_for_id` now:
```cpp
FlEngine* fl_engine_for_id(int64_t id) {
  void* engine = reinterpret_cast<void*>(id);
  g_return_val_if_fail(FL_IS_ENGINE(engine), nullptr);
  return FL_ENGINE(engine);
}
```